test/test_jit.py

```diff
@@ -416,27 +419,65 @@ def f(x, y):
 
         ge = self.checkTrace(f, (x, y))
 
+    @staticmethod
+    def fn_test_relu(x, y):
+        return F.relu(x + .5 * y)
+
     @unittest.skipIf(IS_WINDOWS, "NYI: fuser support for Windows")
     @unittest.skipIf(not RUN_CUDA, "fuser requires CUDA")
     def test_relu(self):
-        def f(x, y):
-            return F.relu(x + .5 * y)
-
         x = torch.randn(4, 4, dtype=torch.float, device='cuda')
         y = torch.randn(4, 4, dtype=torch.float, device='cuda')
 
-        ge = self.checkTrace(f, (x, y))
+        ge = self.checkTrace(self.fn_test_relu, (x, y))
+
+    @staticmethod
+    def fn_test_exp(x, y):
+        return (x + .5 * y).exp()
 
     @unittest.skipIf(IS_WINDOWS, "NYI: fuser support for Windows")
     @unittest.skipIf(not RUN_CUDA, "fuser requires CUDA")
     def test_exp(self):
-        def f(x, y):
-            return (x + .5 * y).exp()
-
         x = torch.randn(4, 4, dtype=torch.float, device='cuda')
         y = torch.randn(4, 4, dtype=torch.float, device='cuda')
 
-        ge = self.checkTrace(f, (x, y))
+        ge = self.checkTrace(self.fn_test_exp, (x, y))
+
+    @unittest.skipIf(IS_WINDOWS, "NYI: fuser support for Windows")
+    @unittest.skipIf(not RUN_CUDA, "fuser requires CUDA")
+    @unittest.skipIf(not RUN_CUDA_HALF, "no half support")
+    def test_cuda_half(self):
+        x = torch.randn(4, 4, dtype=torch.half, device='cuda')
+        y = torch.randn(4, 4, dtype=torch.half, device='cuda')
+
+        funcs = [
+            self.fn_test_comparison_gt_lt,
+            self.fn_test_relu,
+            self.fn_test_exp
+        ]
+
+        # Note: Non fused inputs must be float to prevent loss of precision
+        inputs = (x.float(), y.float())
+        fusion_inputs = (x, y)
+        for fn in funcs:
+            local_inputs = [t.clone().requires_grad_() for t in inputs]
+            local_fusion_inputs = [t.clone().requires_grad_() for t in fusion_inputs]
+
+            # Verifies outputs
+            fusion = torch.jit.trace(*local_fusion_inputs, optimize=True)(fn)
+            outputs = fn(*local_inputs)
+            fusion_outputs = fusion(*local_fusion_inputs)
+            outputs_half = [t.half() for t in outputs]
+            self.assertEqual(outputs_half, fusion_outputs)
```

Comment:
This assert is very confusing for me. If I understand correctly, you simply enabled the half parts of the graph as eligible for fusion, but in here you effectively run the whole graph if `float` (even using `fusion`). I can't see any `half` casts except the `outputs_half` one. Is this really checking the new codepath?

Reply:
I believe you are missing the dtype=torch.half when x and y are created.

The fusion_inputs tuple is composed of half tensors and they create the local_fusion_inputs which are put through the torch.jit.trace call. The inputs tuple is float tensors that are exactly representable in half. 

As I mention above, it would be risky to run halfs through an unfused network because there could be loss of precision within and across operations. That's why I'm doing the work here to compare floats representable as halfs through an unfused network vs the equivalent halfs through the fused network.

Reply:
Ohhh ok. Yes, I fully understand why are we running the unfused thing in float, I just searched for more `half()` calls, but missed the `dtype` part which was too far up for me to see from here. Never mind the comment. LGTM!